[21] Stabilize data graphs

Currently data graphs seem unstable; i.e. are fluctuating between runs.
We need to stabilize them to get consistent results.

- [x] investigate
- [x] fix
- [x] deploy

register list in frozensets?
```
  File "/Users/alterakey/works/trueseeing/wc/.venv/lib/python3.9/site-packages/trueseeing/core/flow/data.py", line 72, in decoded_registers_of_set
    assert False, o
AssertionError: frozenset({'v1', 'v0'})
```

```
  File "/Users/alterakey/works/trueseeing/wc/.venv/lib/python3.9/site-packages/trueseeing/core/flow/data.py", line 72, in decoded_registers_of_set
    assert False, o
AssertionError: frozenset({'v0', 'v1'})
```

--- a/trueseeing/app/shell.py
+++ b/trueseeing/app/shell.py
@@ -159,6 +159,7 @@ Misc:
     for o, a in opts:
       if o in ['-d']:
         log_level = ui.DEBUG
+        ui.is_debugging = True
       if o in ['-o']:
         output_filename = a if a != '-' else None
       if o in ['-W']:

--- a/trueseeing/core/code/model.py
+++ b/trueseeing/core/code/model.py
@@ -40,7 +40,7 @@ class Op:
       self._id = id_
 
   def __repr__(self) -> str:
-    return f'<Op t={self.t} v={self.v}, p={self.p}>'
+    return f'<Op[{self._id}] t={self.t} v={self.v}, p={self.p}>'
 
   def eq(self, t: str, v: str) -> bool:
     return (self.t, self.v) == (t, v)

--- a/trueseeing/core/flow/data.py
+++ b/trueseeing/core/flow/data.py
@@ -24,11 +24,16 @@ from trueseeing.core.flow.code import CodeFlows
 from trueseeing.core.ui import ui
 
 if TYPE_CHECKING:
-  from typing import List, Any, Iterable, Mapping, Set, Optional, FrozenSet, Union
+  from typing import List, Any, Iterable, Mapping, Set, Optional, FrozenSet, Union, Dict
+  from typing_extensions import Final
   from trueseeing.core.store import Store
   from trueseeing.core.code.model import Op
 
+  DataGraph = Union[Op, Mapping[Op, Any]]
+
 class DataFlows:
+  _stash: Final[Dict[int, str]] = dict()
+
   class NoSuchValueError(Exception):
     pass
 
@@ -108,7 +113,7 @@ class DataFlows:
       raise cls.NoSuchValueError(f'not a compile-time constant: {arg!r}')
 
   @classmethod
-  def walk_dict_values(cls, d: Union[Op, Mapping[Op, Any]]) -> Iterable[Optional[Op]]:
+  def walk_dict_values(cls, d: DataGraph) -> Iterable[Optional[Op]]:
     try:
       for v in d.values(): # type: ignore[union-attr]
         yield from cls.walk_dict_values(v)
@@ -149,7 +154,7 @@ class DataFlows:
       return set()
 
   @classmethod
-  def analyze(cls, store: Store, op: Optional[Op], state:Optional[Set[Optional[int]]]=None) -> Optional[Union[Op, Mapping[Op, Any]]]:
+  def analyze(cls, store: Store, op: Optional[Op], state:Optional[Set[Optional[int]]]=None) -> Optional[DataGraph]:
     if op is None:
       return None
     if state is None:

--- a/trueseeing/core/flow/data.py
+++ b/trueseeing/core/flow/data.py
@@ -154,35 +154,51 @@ class DataFlows:
       return set()
 
   @classmethod
-  def analyze(cls, store: Store, op: Optional[Op], state:Optional[Set[Optional[int]]]=None) -> Optional[DataGraph]:
-    if op is None:
+  def analyze(cls, store: Store, op: Optional[Op], state:Optional[Dict[Optional[int], Optional[DataGraph]]]=None, stage:int = 0) -> Optional[DataGraph]:
+    if op is None or stage > 64: # XXX: Is it sufficient? Might be better if we make it tunable?
       return None
     if state is None:
-      state = set()
+      state = dict()
+
+    o: Optional[DataGraph] = None
+
     if op.t == 'id':
       if op._id in state:
-        return None
-      else:
-        state.add(op._id)
+        return state[op._id]
       if any(op.v.startswith(x) for x in ['const','new-','move-exception']):
-        return op
+        o = op
+        state[op._id] = o
+        return o
       elif op.v in ['move', 'array-length']:
-        return {op:{k:cls.analyze(store, cls.analyze_recent_load_of(store, op, k), state) for k in cls.decoded_registers_of_set(op.p[1])}}
+        o = {op:{k:cls.analyze(store, cls.analyze_recent_load_of(store, op, k), state, stage=stage+1) for k in cls.decoded_registers_of_set(op.p[1])}}
+        state[op._id] = o
+        return o
       elif any(op.v.startswith(x) for x in ['aget-']):
         assert len(op.p) == 3
-        return {op:{k:cls.analyze(store, cls.analyze_recent_array_load_of(store, op, k), state) for k in (cls.decoded_registers_of_set(op.p[1]) | cls.decoded_registers_of_set(op.p[2]))}}
+        o = {op:{k:cls.analyze(store, cls.analyze_recent_array_load_of(store, op, k), state, stage=stage+1) for k in (cls.decoded_registers_of_set(op.p[1]) | cls.decoded_registers_of_set(op.p[2]))}}
+        state[op._id] = o
+        return o
       elif any(op.v.startswith(x) for x in ['sget-']):
         assert len(op.p) == 2
-        return {op:{k:cls.analyze(store, cls.analyze_recent_static_load_of(store, op), state) for k in cls.decoded_registers_of_set(op.p[0])}}
+        o = {op:{k:cls.analyze(store, cls.analyze_recent_static_load_of(store, op), state, stage=stage+1) for k in cls.decoded_registers_of_set(op.p[0])}}
+        state[op._id] = o
+        return o
       elif any(op.v.startswith(x) for x in ['iget-']):
         assert len(op.p) == 3
-        return {op:{k:cls.analyze(store, cls.analyze_recent_instance_load_of(store, op), state) for k in cls.decoded_registers_of_set(op.p[0])}}
+        o = {op:{k:cls.analyze(store, cls.analyze_recent_instance_load_of(store, op), state, stage=stage+1) for k in cls.decoded_registers_of_set(op.p[0])}}
+        state[op._id] = o
+        return o
       elif op.v.startswith('move-result'):
-        return cls.analyze(store, cls.analyze_recent_invocation(store, op), state)
+        o = cls.analyze(store, cls.analyze_recent_invocation(store, op), state, stage=stage+1)
+        state[op._id] = o
+        return o
       else:
         try:
-          return {op:{k:cls.analyze(store, cls.analyze_recent_load_of(store, op, k), state) for k in cls.decoded_registers_of_set(op.p[0])}}
+          o = {op:{k:cls.analyze(store, cls.analyze_recent_load_of(store, op, k), state, stage=stage+1) for k in cls.decoded_registers_of_set(op.p[0])}}
+          state[op._id] = o
+          return o
         except cls.RegisterDecodeError:
+          state[op._id] = None
           return None
     else:
       return None

--- a/trueseeing/core/ui.py
+++ b/trueseeing/core/ui.py
@@ -32,6 +32,7 @@ class UI:
   FATAL: Final = 5
 
   level = DEBUG
+  is_debugging = False
 
   def fatal(self, msg: str, nl: bool = True, exc: Optional[Exception] = None) -> NoReturn:
     self.stderr(f'fatal: {msg}', nl=nl, exc=exc)

